### PR TITLE
Add Phi-1 and Phi-1_5

### LINKED ISSUE #26110: [New model] Phi-1 and Phi-1_5
### Model description

From their respective model cards,

phi-1:

> The language model phi-1 is a Transformer with 1.3 billion parameters, specialized for basic Python coding. Its training involved a variety of data sources, including subsets of Python codes from [The Stack v1.2](https://huggingface.co/datasets/bigcode/the-stack), Q&A content from [StackOverflow](https://archive.org/download/stackexchange), competition code from [code_contests](https://github.com/deepmind/code_contests), and synthetic Python textbooks and exercises generated by [gpt-3.5-turbo-0301](https://platform.openai.com/docs/models/gpt-3-5). Even though the model and the datasets are relatively small compared to contemporary Large Language Models (LLMs), phi-1 has demonstrated an impressive accuracy rate exceeding 50% on the simple Python coding benchmark, HumanEval.

phi-1_5:
> The language model phi-1.5 is a Transformer with 1.3 billion parameters. It was trained using the same data sources as [phi-1](https://huggingface.co/microsoft/phi-1), augmented with a new data source that consists of various NLP synthetic texts. When assessed against benchmarks testing common sense, language understanding, and logical reasoning, phi-1.5 demonstrates a nearly state-of-the-art performance among models with less than 10 billion parameters.
>
> We did not fine-tune phi-1.5 either for instruction following or through reinforcement learning from human feedback. The intention behind crafting this open-source model is to provide the research community with a non-restricted small model to explore vital safety challenges, such as reducing toxicity, understanding societal biases, enhancing controllability, and more.

Benchmarks for the 1.5 model (from their [paper](https://arxiv.org/pdf/2309.05463.pdf)) indicate better (or on-par) performance with llama and llama2 models, especially on multi-step reasoning.
image

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Phi-1
Paper: https://arxiv.org/abs/2306.11644

Model + code: https://huggingface.co/microsoft/phi-1

### Phi-1_5

Paper: https://arxiv.org/abs/2309.05463

Model + code: https://huggingface.co/microsoft/phi-1_5

### PR DESCRIPTION
**Pull request by susnato**
# What does this PR do?

Fixes #26110 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

[Colab Link](https://colab.research.google.com/drive/1rY-T36vqG9b-0IM3IJq741zBlcSxSit1?usp=sharing) to verify logits.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc : @xenova

### CODE REVIEW

**Comment by LysandreJik** [comment]
Hey @susnato, thanks for opening your PR! Let us know if we can help in any way.

Also cc @gugarosa and @sgunasekar as you've been doing impressive maintenance work on your models :hugs:

**Comment by susnato** [comment]
Hi @LysandreJik, this PR is ready for review.
Please refer to this [colab notebook](https://colab.research.google.com/drive/1rY-T36vqG9b-0IM3IJq741zBlcSxSit1?usp=sharing) to check the logits matching for `Phi-1` and `Phi-1.5`.

BTW the test - `test_left_padding_compatibility` seems to be failing here, but it is passing on my local machine. Do you know what could be causing this discrepancy?

**Comment by LysandreJik** [comment]
Nice, thanks a lot! Pinging @ArthurZucker and @Rocketknight1 to take a look :)

File: src/transformers/models/phi/modeling_phi.py
Comment (ArthurZucker) [inline src/transformers/models/phi/modeling_phi.py:189]: You should have a look at `PersimmonAtttention` ! It also has a rotary embedding applied to only a portion of the input.

File: src/transformers/models/phi/modeling_phi.py
Comment (ArthurZucker) [inline src/transformers/models/phi/modeling_phi.py:146]: similar to the Persimmon model it seems. Only apply it on `rotary_dim` portion of the input

**Review by ArthurZucker — COMMENTED** [review]
Hey! Seems to me that the modeling is the same as `Persimmon` with just `kq_layer_norm = False` in the config and a different `config.partial_rotary_factor`. 
WDYT?

**Comment by SinclairCoder** [comment]
Will this PR support the training for phi models?

File: src/transformers/models/phi/modeling_phi.py
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py]: yes, updates it in the recent commit.

File: src/transformers/models/phi/modeling_phi.py
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:208]: The attention class for `Phi` is almost same for `Persimmon` except the slight difference in `_split_heads` method. 
I have added `# Copied from` statements to all methods(init, forward, _init_rope) in the attention class except the _split_heads.

**Comment by susnato** [comment]
Hi @SinclairCoder, yes it will.

**Comment by susnato** [comment]
Hello @ArthurZucker, thanks for the suggestions and apologies for the delay.  

I have changed the code quite a bit according to the `persimmon`. For the `PhiAttention`, there is slight difference between the `_split_heads` methods of phi and persimmon so I could not just add `# Copied from` across the whole class, instead I have copied the relevant methods seperately.
In addition to that please note that there are subtle differences in the `PhiDecoderLayer` and `PhiModel` too.

Please review it and let me know if you are ok with the changes or not.

File: src/transformers/models/phi/modeling_phi.py
Comment (ArthurZucker) [inline src/transformers/models/phi/modeling_phi.py:553]: expecting to see a lot more copied from here

**Review by ArthurZucker — COMMENTED** [review]
Thanks for updating ! A few nits but should be close to done

File: src/transformers/models/phi/modeling_phi.py
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:1201]: Yes there are no checkpoints.

File: src/transformers/models/phi/modeling_phi.py
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:282]: > so the subtle difference is the shape that is (3, num_heads) instead of (num_heads, 3)?

yes

Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py]: Done

File: src/transformers/models/phi/modeling_phi.py
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py]: Done

**Comment by susnato** [comment]
Hi @ArthurZucker ,  I have worked on the comments and now `PhiAttention` and `PhiModel` are same as their Persimmon counterparts. 

Please review it and let me know if this works with you.

File: docs/source/en/model_doc/phi.md
Comment (ArthurZucker) [inline docs/source/en/model_doc/phi.md:32]: 🔥

File: src/transformers/models/phi/configuration_phi.py
Comment (ArthurZucker) [inline src/transformers/models/phi/configuration_phi.py:104]:
```suggestion
    model_type = "phi"
```
We are gonna have the same problem as with Faclon, the online model type is `"model_type": "mixformer-sequential",`.
For this, I think we'll have to either keep it or find a workaround! (Otherwise TGI could break)
but that will be the last thing to solve!

File: src/transformers/models/phi/modeling_phi.py
Comment (ArthurZucker) [inline src/transformers/models/phi/modeling_phi.py:560]: its okay to leave that one out 🤣

**Review by ArthurZucker — COMMENTED** [review]
LGTM, I want to approve but we're gonna have a BC compatibility issue with the model type in the config. That's my only concern!

**Comment by ArthurZucker** [comment]
(also make sure the CI are green 🤗 )

File: src/transformers/models/phi/modeling_phi.py
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:560]: Hey we need to keep this line otherwise I am getting repo-consistency error :)

Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py]: Ok

File: docs/source/en/model_doc/phi.md
Comment (susnato) [reply docs/source/en/model_doc/phi.md:72]: Done

**Comment by susnato** [comment]
Hello @ArthurZucker, I have pushed the changes and The CI is green now!

Regarding the BC compatibility issue - 
If it's ok with you, I would like to tackle that step by step after the PR is merged :smiley: .

1. Let's work together to make this PR merged.
2. After it is merged I will create a PR on the huggingface hub to upload both of the weights.
3. After that I will update the checkpoints, demo etc with separate PR just like we did for pop2piano. We can then tackle the compatibility issue.

Please let me know if this works with you :).

File: src/transformers/models/phi/configuration_phi.py
Comment (susnato) [reply src/transformers/models/phi/configuration_phi.py:104]: Can't we just change the `model_type` to `phi` when we update the weights there?

Comment (ArthurZucker) [reply src/transformers/models/phi/configuration_phi.py:104]: No let's just use the one they have online that will be the simplest!

**Comment by ArthurZucker** [comment]
For sure, then we should use the same model_type as them, this way we won't have any problems and should not affect anyone!

**Comment by ArthurZucker** [comment]
(The issue is taht other libraries like TGI use the model_type and otherwise we need to wait for them)

File: src/transformers/models/phi/configuration_phi.py
Comment (susnato) [reply src/transformers/models/phi/configuration_phi.py:104]: Hi @ArthurZucker, I changed the model_type from `"phi"` to `"mixformer-sequential"` but I am getting the `test_load_with_mismatched_shapes` error, I tried to change the mappings in `models/auto` but still getting the error...

Comment (susnato) [reply src/transformers/models/phi/configuration_phi.py]: I have changed `model_type = "phi" -> "mixformer-sequential"` and added a line in the test modelling file `test_mismatched_shapes = False` which skips the failed test for now. 
Later after we transfer the weights and update the online config we can revert back to `model_type = "phi"` and remove the line.

WDYT?

**Comment by susnato** [comment]
Hi @ArthurZucker, I have updated the model_type as theirs(`"mixformer-sequential"`) and mentioned the error I am getting(after the updation) and temporary fix [here](https://github.com/huggingface/transformers/pull/26170#discussion_r1370124238).

File: src/transformers/models/phi/configuration_phi.py
Comment (ArthurZucker) [reply src/transformers/models/phi/configuration_phi.py]: No the error is here to tell you that you need to update the `config_mapin` meaning you need to make sure that `configuration_auto` also has the correct name! 😉 
Let's add a test to make sure that if we load a model from the current revision online with this code , the generated outputs are correct! (using the code_revision and revision flags!)

**Comment by ArthurZucker** [comment]
I am also super down to have the correct model type but we know that's gonna have issue. Last time we had to revert the falcon PR 😓 We don't have the perfect solution yet, I can work on this before we actually push the checkpoints and the config if you are okay to wait? Otherwise we can use their model type for now and later on once I find the correct way to approach this we'll update! 😉

File: src/transformers/models/phi/configuration_phi.py
Comment (susnato) [reply src/transformers/models/phi/configuration_phi.py]: Hey I did change the `CONFIG_MAPPING_NAMES` from `("phi", "PhiConfig")->("mixformer-sequential", "PhiConfig")` But the error was still there :), 
Is there anything more I have to change?

Comment (susnato) [reply src/transformers/models/phi/configuration_phi.py]: Also we can't load either phi-1 or phi-1.5 from the current online version because the weights(namings) won't match :(

**Comment by susnato** [comment]
> I am also super down to have the correct model type but we know that's gonna have issue. Last time we had to revert the falcon PR 😓 We don't have the perfect solution yet, I can work on this before we actually push the checkpoints and the config if you are okay to wait?

If you think that's relatively better solution, then I will wait. 

But then could you please explain me the whole process a little bit?

Also I have a question - after the PR is merged, we will need to delete the modelling files on the HuggingFace Hub right? So that everyone loads the model from the library only?

File: src/transformers/models/phi/configuration_phi.py
Comment (susnato) [reply src/transformers/models/phi/configuration_phi.py]: Please see [this comment](https://github.com/huggingface/transformers/pull/26170#issuecomment-1778720835).

**Comment by bnicholl** [comment]
when importing phi with transformers == 4.34.1 in the way the below docs describe. `from transformers import PhiModel`
[](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26170/en/model_doc/phi)

I get an error: ImportError: cannot import name 'PhiForCausalLM'

**Comment by ArthurZucker** [comment]
Hey @susnato, we'll probably merge this PR, then open a PR to the repo on the hub with the new checkpoints and a deletion of the code, making sure checking out this revision works, then merging once the release happens (next week). I'll let @Rocketknight1 help you with this as he went thought that for Falcon! 

Regarding this PR, let's merge it but with the correct model_type (so the one they have in their config)  for now. I didn't have time to dive into fixing and will be off for a week so will tackle after that!

**Comment by susnato** [comment]
Thanks for the reply @ArthurZucker !

**Comment by susnato** [comment]
Hey @ArthurZucker, I tried to change to `mixformer-sequential` for `MODEL_NAMES_MAPPING` and `CONFIG_MAPPING_NAMES` but it seems that I need to change the whole directory from `phi` to `mixformer-sequential` in order to make it work.
 

Otherwise I believe that we can also add  `("mixformer-sequential", "phi")` to `SPECIAL_MODEL_TYPE_TO_MODULE_NAME` so that every `mixformer-sequential` will get converted to `phi`.

WDYT?

**Comment by ArthurZucker** [comment]
Mmm yeah something like `HUB_MODEL_TYPE_TO_TRANSFORMERS_MODEL_TYPE` will have to be created, but it's a bit more involved we need to test and make sure we handle edge case

**Comment by ArthurZucker** [comment]
I'll try to do this today 👍🏻

**Comment by susnato** [comment]
So what do we do now? Do we merge with model_type as phi? or keep the mixformer-sequential(but override the case as False)?

**Comment by Rocketknight1** [comment]
Just replying here to say that yes, I'm available to help when you're ready to start porting checkpoints!

**Comment by susnato** [comment]
Hello @Rocketknight1, thanks for the reply!

I have already ported the checkpoints and saved here - [phi-1](https://huggingface.co/susnato/phi-1_dev) and [phi-1.5](https://huggingface.co/susnato/phi-1_5_dev)

The main [issue](https://github.com/huggingface/transformers/pull/26170#pullrequestreview-1692913227) that @ArthurZucker was talking about was the difference between the `model_type`, the [online config file, on the hub](https://huggingface.co/microsoft/phi-1/blob/371fd519ab41b5bc2e96838936c2349dc49656db/config.json#L14) has model_type as `mixformer-sequential` but here we have it as `phi`. This won't matter if we change the model_type to phi on the Hub.
Also I believe that we will need to delete the modelling files on the HuggingFace Hub right?

So, how should I approach this? should I first open a PR to transfer the weights and config to the original repo on Hub or this PR needs to be merged first then the weights transfer? WDYT?

**Comment by Rocketknight1** [comment]
@susnato This PR will need to be merged first, because otherwise the updated repos won't work for anyone else.

After that, the way we did it was to make a PR to each checkpoint repo, and do the following:

1) Copy the `configuration_phi.py` and `modeling_phi.py` files from this PR into the repo, overwriting the old ones.
2) Ensure the files don't use relative imports (e.g. replace `from ..modeling_utils import` with `from transformers.modeling_utils import`
3) Set the `architectures` and `auto_map` keys in `config.json` to point to the new classes.
4) Set the `model_type` to the new type (e.g. `phi` in this case)

This way, users who have updated to the latest version of `transformers` can use `trust_remote_code=False` and get the library code, but users who haven't can continue with their old workflows.

**Comment by Rocketknight1** [comment]
@susnato we should use "phi"

**Comment by susnato** [comment]
@Rocketknight1, Okk. Then lets merge this first.

If you don't mind could you please review this? :sweat_smile:

**Comment by Rocketknight1** [comment]
@susnato will do! Please do the rebase anyway though, lol. I think we'll get lucky, and if we don't then it's better to find out here then after we merge!

File: src/transformers/models/phi/modeling_phi.py
Comment (Rocketknight1) [inline src/transformers/models/phi/modeling_phi.py:87]: This isn't a huge problem, but for new models we prefer to use [AttentionMaskConverter._expand_mask()](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_attn_mask_utils.py#L150) for this! It's exactly the same code, it just avoids everyone copying this one function from BART.

**Review by Rocketknight1 — APPROVED** [review]
Given the similarities with `Persimmon`, this looks like a pretty clean port without a lot of deep code to review, so I'm happy to accept it!

I'd like two things before we merge, though:

1) The rebase (sorry!)
2) To double-check with Arthur and the TGI team about the model name. I'm 99% sure `phi` is acceptable here, but I want to confirm all of that before we merge. I'll get this done by tomorrow!

File: src/transformers/models/phi/modeling_phi.py
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:87]: Ok

Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py]: Done

**Comment by susnato** [comment]
Hello @Rocketknight1, I have rebased and did necessary changes.

Please note that I have also added an [dropout layer](https://github.com/susnato/transformers/blob/26cceddd5504d5f58d899469483a0245f082f0ee/src/transformers/models/phi/modeling_phi.py#L610) after the `input_embeds` which was missing in the previous commits.

**Comment by Rocketknight1** [comment]
Got it! Pinging either @arthurzucker or @amyeroberts for core maintainer final review. I checked with the TGI team and they said it would be fine to do the port and change the model type to `phi`.

File: docs/source/en/model_doc/phi.md
Comment (amyeroberts) [inline docs/source/en/model_doc/phi.md:93]: So would I 🥲

File: docs/source/en/model_doc/phi.md
Comment (amyeroberts) [inline docs/source/en/model_doc/phi.md:1]: There was a recent [push to main](https://github.com/huggingface/transformers/commit/5964f820db1568d26298b37dea9db328185c7f7c) which included standardisation of our model doc page structures and section titles. Could you update this to reflect those e.g. using `Usage examples`, and putting tips inside an embedded `<Tip>` section

File: src/transformers/models/phi/modeling_phi.py
Comment (amyeroberts) [inline src/transformers/models/phi/modeling_phi.py:754]: Let's avoid example where we talk about consciousness for models

**Review by amyeroberts — COMMENTED** [review]
Awesome work - thanks for adding this model! 

Just some nits on documentation and examples. Once updated we'll be good to go!

File: docs/source/en/model_doc/phi.md
Comment (susnato) [reply docs/source/en/model_doc/phi.md:93]: But it might have different motives!! :laughing:

File: src/transformers/models/phi/modeling_phi.py
Comment (bnicholl) [reply src/transformers/models/phi/modeling_phi.py:754]: In the above example we use 
`model = PhiForCausalLM.from_pretrained("susnato/phi-1_dev")`, where PhiForCausalLM gives an import error.

In the official docs we're using 
`model = AutoModelForCausalLM.from_pretrained("microsoft/phi-1_5", trust_remote_code=True)`.

Are both of these functions of loading the model the same? I only ask because I know with previous versions, you could not fine tune the model, so I wanna make sure fine tuning is now corrected using both function calls.

Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:754]: Hi @bnicholl, if you want to use `PhiForCausalLM` right now you need to checkout my branch (`susnato:phi`) and then install from there, otherwise wait for this PR to get merged. :)

> Are both of these functions of loading the model the same? I only ask because I know with previous versions, you could not fine tune the model, so I wanna make sure fine tuning is now corrected using both function calls.

Once this PR is merged it will load the same model and yes you can fine-tune like any other models from this library!

**Comment by susnato** [comment]
Hi @amyeroberts, I have reworked the model doc page. Let me know if this is okay to you.

**Review by amyeroberts — APPROVED** [review]
Thanks again for adding and iterating!

**Comment by susnato** [comment]
Looks like the CI error is unrelated to this PR. Should I rebase and push again? @amyeroberts

**Comment by amyeroberts** [comment]
@susnato Yes please! Sorry - there were a few unexpected error because of new packages releases and incompatibilities. These should now be resolved on main.

**Comment by susnato** [comment]
Hi, @amyeroberts I just rebased but the failing test seems to be unrelated.

**Comment by amyeroberts** [comment]
@susnato Indeed. Really sorry about this, we're trying to resolve now. It seems there's some flakiness in accelerate tests - which means even if we _think_ we've fixed them another rears its head. I'll let you know once the latest patch is merged. Fingers crossed that will be the last rebase you'll have to do.

**Comment by susnato** [comment]
No problem @amyeroberts, just ping me when it's patched and I will rebase then.

**Comment by amyeroberts** [comment]
OK, third time lucky. Apologies for the delay in getting this merged and thank you for your patience. A fix which has been run across the full test suite has been merged into main. If you rebase this _should_ resolve the issue

**Comment by susnato** [comment]
Hmm it failed on another test :( but it was ok for the CLVP PR,  let me rebase and push one more time.

**Comment by susnato** [comment]
Hi @amyeroberts, the CI is green now!

**Comment by amyeroberts** [comment]
@susnato Amazing! Thanks for your work adding this and your patience whilst we sorted out the CI

**Comment by Rocketknight1** [comment]
Hi @susnato, congrats on the PR! Let me know if you're willing to tackle opening PRs to the model repos as well. I mentioned the recommended approach [here](https://github.com/huggingface/transformers/pull/26170#issuecomment-1787585303).

**Comment by susnato** [comment]
Hi @Rocketknight1, yes I will open PR to both of the repos on the Hub this week.

**Comment by Rocketknight1** [comment]
Amazing, thank you!

**Comment by susnato** [comment]
Hi @Rocketknight1, sorry I could not open the PRs this week, I will try to complete them ASAP.

**Comment by susnato** [comment]
Hi @Rocketknight1, I have created both the PRs - [phi-1](https://huggingface.co/microsoft/phi-1/discussions/5) and [phi-1.5](https://huggingface.co/microsoft/phi-1_5/discussions/62).

Could you please review them?

**Comment by liziniu** [comment]
Hi all,

It seems that the current code for Phi-1.5 does not support gradient checkpointing. Could you support this?

**Comment by susnato** [comment]
Hi @liziniu, gradient checkpointing is already supported by `Phi`.

```python
from transformers import PhiForCausalLM

model = PhiForCausalLM.from_pretrained("susnato/phi-1_5_dev") # the weights are currently being transferred to microsoft/phi-1_5
print(model.model.gradient_checkpointing) # False

# enable Gradient Checkpointing
model.gradient_checkpointing_enable()

print(model.model.gradient_checkpointing) # this should be True now.
```

**Comment by mohamedjihedriahi** [comment]
Hello folks

I've encountered an issue while attempting to load my fine-tuned model after the recent update. The error message I received is as follows:

"OSError: microsoft/phi-1_5 does not seem to contain a file named configuration_mixformer_sequential.py. Please check 'https://huggingface.co/microsoft/phi-1_5/main' for available files."

As someone relatively new to this, can anyone help me figure out how to tweak my code so I can load my fine-tuned model properly?
```
model = AutoModelForCausalLM.from_pretrained("nulltella/phi-1_5-finetuned-model", trust_remote_code=True, torch_dtype=torch.float32)
tokenizer = AutoTokenizer.from_pretrained("microsoft/phi-1_5", trust_remote_code=True)
```
Thanxxx !
#Help

**Comment by susnato** [comment]
Hi @mohamedjihedriahi, as far as I can tell from the error message, your code is trying to find `configuration_mixformer_sequential.py` but that file was deleted in this [commit](https://huggingface.co/microsoft/phi-1_5/commit/8e9ebfb9bfe4da87d5c705d097fe884dc8e62d6f). So you are getting this error.


**Short Time Fix**
For a quick fix, you can try to convert your `nulltella/phi-1_5-finetuned-model` to have the same `config.json` as `microsoft/phi-1_5` according to [this commit](https://huggingface.co/microsoft/phi-1_5/commit/271c3397ab4e1f8f4e49868b1e8ba0be95363c88).

**Long Time Fix (Recommended)**
Since `phi` is now added to `transformers`, I would suggest you - 
1. upgrade to the latest `transformers`.
2. convert the previous weights according to the library model's weight structure. (this would require you to rearrange the keys of the weights).

This way, you won't have to care about any new changes to the model `phi` in the library. 


Let me know if you need any assistance/help in doing this.

**Pull request merged**
